src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java

```diff
@@ -181,23 +219,147 @@ public Set<Component> getComponents()
         return getComponents(indexes);
     }
 
-    // TODO: Currently returns an empty set for in-memory only implementation
-    static Set<Component> getComponents(Collection<Index> indices)
+    static Set<Component> getComponents(Collection<StorageAttachedIndex> indices)
+    {
+        Set<Component> components = Version.LATEST.onDiskFormat()
+                                                  .perSSTableComponents()
+                                                  .stream()
+                                                  .map(Version.LATEST::makePerSSTableComponent)
+                                                  .collect(Collectors.toSet());
+        indices.forEach(index -> components.addAll(index.getComponents()));
+        return components;
+    }
+
+    // This differs from getComponents in that it only returns index components that exist on disk.
+    // It avoids errors being logged by the SSTable.readTOC method when we have an empty index.
+    @VisibleForTesting
+    public static Set<Component> getLiveComponents(SSTableReader sstable, Collection<StorageAttachedIndex> indices)
     {
-        return Collections.emptySet();
+        IndexDescriptor indexDescriptor = IndexDescriptor.create(sstable);
+        Set<Component> components = indexDescriptor.getLivePerSSTableComponents();
+        indices.forEach(index -> components.addAll(indexDescriptor.getLivePerIndexComponents(index.getIndexContext())));
+        return components;
     }
 
     @Override
     public void handleNotification(INotification notification, Object sender)
     {
-        if (notification instanceof MemtableRenewedNotification)
+        // unfortunately, we can only check the type of notification via instanceof :(
+        if (notification instanceof SSTableAddedNotification)
         {
-            indexes.forEach(index -> ((StorageAttachedIndex)index).getIndexContext().renewMemtable(((MemtableRenewedNotification) notification).renewed));
+            SSTableAddedNotification notice = (SSTableAddedNotification) notification;
+
+            // Avoid validation for index files just written following Memtable flush.
+            boolean validate = !notice.memtable().isPresent();
+            onSSTableChanged(Collections.emptySet(), notice.added, indexes, validate);
+        }
+        else if (notification instanceof SSTableListChangedNotification)
+        {
+            SSTableListChangedNotification notice = (SSTableListChangedNotification) notification;
+
+            // Avoid validation for index files just written during compaction.
+            onSSTableChanged(notice.removed, notice.added, indexes, false);
+        }
+        else if (notification instanceof MemtableRenewedNotification)
+        {
+            indexes.forEach(index -> index.getIndexContext().getMemtableIndexManager().renewMemtable(((MemtableRenewedNotification) notification).renewed));
```

Comment:
nit: In places like this, perhaps we could pull out the `getMemtableIndexManager()` call just to save a few assertions?

Reply:
We can't do that because each index has it's own `MemtableIndexManager`. I understand the concern about the asserts but I would prefer to address that by correctly decoupling the indexed and non-indexed contexts with CASSANDRA-18166.